pkg/vault/contracts/RouterHooks.sol

```diff
@@ -0,0 +1,402 @@
+// SPDX-License-Identifier: GPL-3.0-or-later
+
+pragma solidity ^0.8.24;
+
+import { IERC4626 } from "@openzeppelin/contracts/interfaces/IERC4626.sol";
+import { SafeCast } from "@openzeppelin/contracts/utils/math/SafeCast.sol";
+import { IERC20 } from "@openzeppelin/contracts/token/ERC20/IERC20.sol";
+import { Address } from "@openzeppelin/contracts/utils/Address.sol";
+import { IPermit2 } from "permit2/src/interfaces/IPermit2.sol";
+
+import { IWETH } from "@balancer-labs/v3-interfaces/contracts/solidity-utils/misc/IWETH.sol";
+import { IVault } from "@balancer-labs/v3-interfaces/contracts/vault/IVault.sol";
+import "@balancer-labs/v3-interfaces/contracts/vault/VaultTypes.sol";
+import "@balancer-labs/v3-interfaces/contracts/vault/RouterTypes.sol";
+
+import { RouterWethLib } from "./lib/RouterWethLib.sol";
+import { RouterCommon } from "./RouterCommon.sol";
+
+/**
+ * @notice Base Router with hooks for swaps and liquidity operations via Vault.
+ * @dev Implements hooks for init, add liquidity, remove liquidity, and swaps.
+ */
+contract RouterHooks is RouterCommon {
+    using Address for address payable;
+    using RouterWethLib for IWETH;
+    using SafeCast for *;
+
+    constructor(
+        IVault vault,
+        IWETH weth,
+        IPermit2 permit2,
+        string memory routerVersion
+    ) RouterCommon(vault, weth, permit2, routerVersion) {
+        // solhint-disable-previous-line no-empty-blocks
+    }
+
+    /**
+     * @notice Hook for initialization.
+     * @dev Can only be called by the Vault.
+     * @param params Initialization parameters (see IRouter for struct definition)
+     * @return bptAmountOut BPT amount minted in exchange for the input tokens
+     */
+    function initializeHook(
+        InitializeHookParams calldata params
+    ) external nonReentrant onlyVault returns (uint256 bptAmountOut) {
+        return _initializeHook(params);
+    }
+
+    function _initializeHook(InitializeHookParams calldata params) internal returns (uint256 bptAmountOut) {
+        bptAmountOut = _vault.initialize(
+            params.pool,
+            params.sender,
+            params.tokens,
+            params.exactAmountsIn,
+            params.minBptAmountOut,
+            params.userData
+        );
+
+        for (uint256 i = 0; i < params.tokens.length; ++i) {
+            IERC20 token = params.tokens[i];
```

Comment:
Could move the token assignment lower so that it's not executed in the "continue" case.

Reply:
note that we're not doing any changes to the router logic in this PR. This is just copying and re-arranging existing logic to make the review easier.